### PR TITLE
CalendarEntryAttachment Service Add

### PR DESCRIPTION
**Pull request by viralpandya**

### CODE REVIEW

File: OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Exceptions.Add.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            // given
+            CalendarEntryAttachment randomCalendarEntryAttachment = CreateRandomCalendarEntryAttachment();
+            CalendarEntryAttachment inputCalendarEntryAttachment = randomCalendarEntryAttachment;
+            var sqlException = GetSqlException();
+            
+            var expectedCalendarEntryAttachmentDependencyException = 
+                new CalendarEntryAttachmentDependencyException(sqlException);
+
+            this.storageBrokerMock.Setup(broker =>
+                broker.InsertCalendarEntryAttachmentAsync(inputCalendarEntryAttachment))
```
Comment (hassanhabib) [inline OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Exceptions.Add.cs:30]: We don't really care what the input is

File: OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Exceptions.Add.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            // when
+            ValueTask<CalendarEntryAttachment> addCalendarEntryAttachmentTask =
+                this.calendarEntryAttachmentService.AddCalendarEntryAttachmentAsync(inputCalendarEntryAttachment);
+
+            // then
+            await Assert.ThrowsAsync<CalendarEntryAttachmentDependencyException>(() =>
+                addCalendarEntryAttachmentTask.AsTask());
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(expectedCalendarEntryAttachmentDependencyException))),
```
Comment (hassanhabib) [inline OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Exceptions.Add.cs:78]: Logging should come after the call verification

File: OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Validations.Add.cs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+namespace OtripleS.Web.Api.Tests.Unit.Services.CalendarEntryAttachments
+{
+    public partial class CalendarEntryAttachmentServiceTests
+    {
+        [Fact]
+        public async void ShouldThrowValidationExceptionOnAddWhenCalendarEntryAttachmentIsNullAndLogItAsync()
+        {
+            // given
+            CalendarEntryAttachment randomCalendarEntryAttachment = default;
```
Comment (hassanhabib) [inline OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Validations.Add.cs:21]: This is not random its null

File: OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Exceptions.Add.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+namespace OtripleS.Web.Api.Tests.Unit.Services.CalendarEntryAttachments
+{
+    public partial class CalendarEntryAttachmentServiceTests
+    {
+        [Fact]
+        public async Task ShouldThrowDependencyExceptionOnAddWhenSqlExceptionOccursAndLogItAsync()
+        {
+            // given
+            CalendarEntryAttachment randomCalendarEntryAttachment = CreateRandomCalendarEntryAttachment();
+            CalendarEntryAttachment inputCalendarEntryAttachment = randomCalendarEntryAttachment;
```
Comment (hassanhabib) [inline OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Exceptions.Add.cs:23]: We just say some Entity

File: OtripleS.Web.Api/Services/CalendarEntryAttachments/CalendarEntryAttachmentService.cs [hunk 5]
```diff
@@ -28,6 +28,15 @@ public partial class CalendarEntryAttachmentService : ICalendarEntryAttachmentSe
             this.dateTimeBroker = dateTimeBroker;
         }
 
+        public ValueTask<CalendarEntryAttachment> AddCalendarEntryAttachmentAsync(
+            CalendarEntryAttachment calendarEntryAttachment) =>
+        TryCatch(async () =>
```
Comment (hassanhabib) [inline OtripleS.Web.Api/Services/CalendarEntryAttachments/CalendarEntryAttachmentService.cs:33]: Move this up by one

File: OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Validations.Add.cs [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async void ShouldThrowValidationExceptionOnAddWhenReferneceExceptionAndLogItAsync()
+        {
+            // given
+            CalendarEntryAttachment randomCalendarEntryAttachment = CreateRandomCalendarEntryAttachment();
+            CalendarEntryAttachment invalidCalendarEntryAttachment = randomCalendarEntryAttachment;
+            string randomMessage = GetRandomMessage();
```
Comment (hassanhabib) [inline OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Validations.Add.cs:167]: This variable is not invalid. we also don't really care about the variable here. it's just a fill-in. it should be just called `someCalendarEntryAttachment`

File: OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Validations.Add.cs [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+            var foreignKeyConstraintConflictException = new ForeignKeyConstraintConflictException(exceptionMessage);
+
+            var invalidCalendarEntryAttachmentReferenceException =
+                new InvalidCalendarEntryAttachmentReferenceException(foreignKeyConstraintConflictException);
+
+            var expectedCalendarEntryAttachmentValidationException =
+                new CalendarEntryAttachmentValidationException(invalidCalendarEntryAttachmentReferenceException);
+
+            this.storageBrokerMock.Setup(broker =>
+                broker.InsertCalendarEntryAttachmentAsync(invalidCalendarEntryAttachment))
```
Comment (hassanhabib) [inline OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Validations.Add.cs:178]: The input here should be `It.IsAny<CalendarEntryAttachment>`

Comment (hassanhabib) [reply OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Validations.Add.cs:178]: Do the same with the rest.

File: OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Validations.Add.cs (hunk 6 above)
Comment (hassanhabib) [reply OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Validations.Add.cs:167]: Do the same with the rest

Comment (viralpandya) [reply OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Validations.Add.cs:167]: These are validation testcases and this was written for InvalidCalendarEntryAttachmentReferenceException. if error is related to input then it should be invalid..... But if error is not related to input then we name input object as some...

Please clarify this.

File: OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Exceptions.Add.cs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+namespace OtripleS.Web.Api.Tests.Unit.Services.CalendarEntryAttachments
+{
+    public partial class CalendarEntryAttachmentServiceTests
+    {
+        [Fact]
+        public async Task ShouldThrowDependencyExceptionOnAddWhenSqlExceptionOccursAndLogItAsync()
+        {
+            // given
+            CalendarEntryAttachment randomCalendarEntryAttachment = CreateRandomCalendarEntryAttachment();
```
Comment (hassanhabib) [inline OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Exceptions.Add.cs:22]: You don't need this line completely when you're creating `someVariable` so you should say:

```csharp
 CalendarEntryAttachment someCalendarEntryAttachment= CreateRandomCalendarEntryAttachment();
```

Comment (hassanhabib) [reply OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Exceptions.Add.cs:22]: Do the same for the rest.

Comment (viralpandya) [reply OtripleS.Web.Api.Tests.Unit/Services/CalendarEntryAttachments/CalendarEntryAttachmentServiceTests.Exceptions.Add.cs]: Resolved in all exceptions for Add, Remove and RetrieveById

**Pull request merged**